apps/src/lib/node/ledger/shell/mod.rs

```diff
@@ -765,7 +765,12 @@ where
                         if let Err(err) = self
                             .validate_valset_upd_vext_and_get_it_back(
                                 ext,
-                                self.storage.last_height,
+                                // n.b. only accept validator set updates
+                                // issued at
+                                // the current epoch (signing off on the
+                                // validators
+                                // of the next epoch)
+                                self.storage.get_current_epoch().0,
```

Comment:
`self.storage.get_current_epoch()` returns different values depending on the phase of ABCI++ we are at (i.e. if we are at `FinalizeBlock` or `Commit`, it will be the epoch of the block that's not yet committed, but outside of those phases, it's the epoch of the most recently committed block) - would that matter here? 

More generally, we need a better API around getting "current" block heights I think (maybe a state machine for tracking which phase of ABCI++ we are in, and implementing methods differently for certain states? e.g. `Storage<FinalizeBlock|Commit>.get_current_epoch()` and `Storage<OtherPhases>.get_current_epoch()`

Reply:
a type level state machine wouldn't work, but an enum with the current phase is possible

Reply:
If we can, let's use `storage.last_height` (i.e. most recently committed block height) or something derived from it, as that value doesn't depend on which phase of ABCI we're in